src/crypto.rs

```diff
@@ -0,0 +1,206 @@
+    fn set_property_rc2_effective_key_bits(&self, rc2_eff_key_bits: usize) -> Result<()>
+    {
+        context_set_property_single(self, types::Property::Rc2EffectiveKeyBits, rc2_eff_key_bits)
+    }
+}
+
+/// Implementation of GCM/CCM set/get properties for encrypt/seal
+pub trait ContextWithXcmEncryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag length property.
+    fn set_property_gcm_tag_len(&self, gcm_tag_len: usize) -> Result<()>
+    {
+        context_set_property_single(self, types::Property::GcmTagLen, gcm_tag_len)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag length property.
+    fn set_property_ccm_tag_len(&self, ccm_tag_len: usize) -> Result<()>
+    {
+        context_set_property_single(self, types::Property::CcmTagLen, ccm_tag_len)
+    }
+    /// Returns the GCM tag property.
+    fn get_property_gcm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::GcmTag)
+    }
+    /// Returns the CCM tag property.
+    fn get_property_ccm_tag(&self) -> Result<Vec<u8>>
+    {
+        context_get_property_multiple(self, types::Property::CcmTag)
+    }
+}
+
+/// Implementation of all GCM/CCM properties for decrypt/open
+pub trait ContextWithXcmDecryptProperties: Context {
+    /// Sets the GCM AAD property.
+    fn set_property_gcm_aad(&self, gcm_aad: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmAad, gcm_aad)
+    }
+    /// Sets the GCM tag property.
+    fn set_property_gcm_tag(&self, gcm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::GcmTag, gcm_tag)
+    }
+    /// Sets the CCM AAD property.
+    fn set_property_ccm_aad(&self, ccm_aad: &[u8], input_len: usize) -> Result<()>;
+    // {
+    //     OPERATION_set_input_length(self, input_len)?;
+    //     crypto::context_set_property_multiple(ctx, types::Property::CcmAad, ccm_aad)
+    // }
+    /// Sets the CCM tag property.
+    fn set_property_ccm_tag(&self, ccm_tag: &[u8]) -> Result<()>
+    {
+        context_set_property_multiple(self, types::Property::CcmTag, ccm_tag)
+    }
+}
+
+
+/// Initializes the library. Must be called before any other crypto
+/// function. Should be called once in each thread that uses yaca.
+pub fn initialize() -> Result<()>
+{
+    let r = unsafe {
+        lib::yaca_initialize()
+    };
+    conv::res_c_to_rs(r)
+}
+
+/// Cleans up the library.
+/// Must be called before exiting the thread that called yaca_initialize().
+pub fn cleanup()
```

Comment:
Initialiaze was supposed to be called per every thread afair. This would cause a confusion.

Reply:
Ok. Would having ```initialize``` produce a token and ```cleanup``` consume the token be reasonable then?

Reply:
What for? And how would you see this implemented? They'd have to match? Where would I store it on the yaca side?

Reply:
Purely rust-side, an empty struct, no matching required, just to leverage borrowck to ensure at most one ```cleanup``` per initialize.

Reply:
1. This would make any sense if that empty struct was not Send + Sync. Otherwise you'd still be able to call 2 cleanups per initialize (at least one of them)
2. More then one cleanup per thread is not forbidden. This solves nothing.

Reply:
How about a ```cleanup``` with no prior ```initialize```?

Reply:
noop as well. Look at the C code.

Reply:
My remark is not exactly about what it ends up doing in the end but about whether we should leverage borrowck to discourage nonsensical use patterns. Regardless, I get where you're coming from.

Reply:
Is there any way to make an empty struct not impl Send and Sync?

Reply:
```#![feature(negative_impls)]``` is unstable unfortunately. Or hide a dummy raw pointer inside.